crates/starknet-os/src/io/output.rs

```diff
@@ -134,99 +134,140 @@ fn next_as_usize<I: Iterator<Item = Felt252>>(output_iter: &mut I, item_name: &s
 
-    for i in 0..n_class_changes {
+    for i in 0..n_actual_updates {
         let class_hash = next_or_fail(output_iter, &format!("class hash #{i}"))?;
+        if !full_output.is_zero() {
+            next_or_fail(output_iter, &format!("previous compiled class hash #{i}"))?;
+        }
         let compiled_class_hash = next_or_fail(output_iter, &format!("compiled class hash #{i}"))?;
-
         classes.insert(class_hash, compiled_class_hash);
     }
 
     Ok(classes)
 }
 
-pub fn decode_output<I: Iterator<Item = Felt252>>(output_iter: &mut I) -> Result<StarknetOsOutput, SnOsError> {
+fn deserialize_messages<I>(output_iter: &mut I) -> Result<(Vec<Felt252>, Vec<Felt252>), SnOsError>
+where
+    I: Iterator<Item = Felt252>,
+{
     /// Reads a section with a variable length from the iterator.
     /// Some sections start with a length field N followed by N items.
     fn read_variable_length_segment<I: Iterator<Item = Felt252>>(
         output_iter: &mut I,
         item_name: &str,
     ) -> Result<Vec<Felt252>, SnOsError> {
         let n_items = next_as_usize(output_iter, item_name)?;
-        let items = output_iter.by_ref().take(n_items).collect();
-
-        Ok(items)
+        read_segment(output_iter, n_items, item_name)
     }
 
-    let header: Vec<Felt252> = output_iter.by_ref().take(HEADER_SIZE).collect();
-    if header.len() != HEADER_SIZE {
+    let messages_to_l1 = read_variable_length_segment(output_iter, "L1 messages")?;
+    let messages_to_l2 = read_variable_length_segment(output_iter, "L2 messages")?;
+    Ok((messages_to_l1, messages_to_l2))
+}
+
+fn read_segment<I: Iterator<Item = Felt252>>(
+    output_iter: &mut I,
+    length: usize,
+    item_name: &str,
+) -> Result<Vec<Felt252>, SnOsError> {
+    let segment = output_iter.by_ref().take(length).collect::<Vec<_>>();
+    if segment.len() != length {
         return Err(SnOsError::CatchAll(format!(
-            "Expected the header to have {} elements, could only read {}",
-            HEADER_SIZE,
-            header.len()
+            "Expected {} {}, could only read {}",
+            length,
+            item_name,
+            segment.len()
         )));
     }
+    Ok(segment)
+}
 
-    let use_kzg_da = {
-        let use_kzg_da_felt = header[USE_KZG_DA_OFFSET];
-        if use_kzg_da_felt == Felt252::ZERO {
-            false
-        } else if use_kzg_da_felt == Felt252::ONE {
-            true
-        } else {
-            return Err(SnOsError::CatchAll(format!("Invalid KZG flag: {}", use_kzg_da_felt.to_biguint())));
-        }
-    };
+pub fn deserialize_os_output<I>(output_iter: &mut I) -> Result<StarknetOsOutput, SnOsError>
+where
+    I: Iterator<Item = Felt252>,
+{
+    let header = read_segment(output_iter, HEADER_SIZE, "header elements")?;
+    let use_kzg_da = header[USE_KZG_DA_OFFSET];
+    let full_output = header[FULL_OUTPUT_OFFSET];
 
-    if use_kzg_da {
+    if !use_kzg_da.is_zero() {
```

Comment:
What is the difference between this and just `use_kzg_da`?

Reply:
`use_kzg_da` used to be `bool` converted from the Felt. I've kept the original Felt value and done a zero check, mirroring what the Cairo code is doing.